### DISCUSSION TITLE
Bug: Nested rule names don't work with flat config

### DISCUSSION THREAD

**Discussion by meeroslav** [eslint/eslint#17901] [Q&A]
### Environment

Node version: v20
npm version: v10
Local ESLint version: 8.48.0
Global ESLint version: -
Operating System: MacOS Ventura (M1)

### What parser are you using?

Default (Espree)

### What did you do?

We have an ESlint plugin that dynamically adds users' rules as a nested rules using the convention `@nx/workspace/{ruleName}`. All the other static rules follow the typical convention `@nx/{ruleName}`.

You can see the code for it here:
https://github.com/nrwl/nx/blob/master/packages/eslint-plugin/src/index.ts#L40

### What did you expect to happen?

This works fine with old configs/CLI.

### What actually happened?

Running the same plugin with flat configs and environment flag returns following error:

```
TypeError: Key "rules": Key "@nx/workspace/demo-rule": Could not find plugin "@nx/workspace".
```

### Link to Minimal Reproducible Example

https://github.com/beaussan/repro-nx-workspace-rule-with-flat-config

### Participation

- [X] I am willing to submit a pull request for this issue.

### Additional comments

Changing resolution logic to name dynamic rules like `@nx/workspace-{ruleName}` works. Is this accidental omission on ESLint's side or are the nested rules names not supported anymore?

**Comment by meeroslav** [comment]
I realized this is by design https://github.com/eslint/eslint/blob/main/lib/config/flat-config-helpers.js#L25-L29 to support plugin names such as `@scope/name`. Not sure how/why this works in non-flat config.

Additionally, the logic above seems a bit odd as it converts string like `@scope/foo/bar/baz` into:
```
{ pluginName: '@scope/foo/bar', ruleName: 'baz' }
```

How can can plugin be named `@scope/foo/bar`? I would expect the logic to be something like:
```
        if (ruleId.startsWith("@")) {
            pluginName = ruleId.slice(0, ruleId.indexOf("/", ruleId.indexOf("/") + 1));
        } else {
            pluginName = ruleId.slice(0, ruleId.indexOf("/"));
        }

```

which would map above string to:
```
{ pluginName: '@scope/foo', ruleName: 'bar/baz' }
```

**Comment by nzakas** [comment]
Yes, you can't have a plugin namespace of `@nx`, as in your example, as this isn't a valid plugin name in eslintrc. We carried that behavior over for compatibility reasons. I'd expect you to still use `@nx/eslint-plugin` as the plugin namespace.

**Comment by meeroslav** [comment]
Yes, we use the `@nx/eslint-plugin` so each rule is scoped under `@nx`:
- @nx/enforce-module-boundaries
- @nx/dependency-checks

And I have renamed the old `@nx/workspace/{rule}` to `@nx/workspace-{rule}`. This works fine now in both classic and flat config.

Can you comment on my second message @nzakas? How come is `@scope/foo/bar/baz` parsed to:
```
{ pluginName: '@scope/foo/bar', ruleName: 'baz' }
```

The tests explicitly check this, so it must be intentional. That 3-segment name doesn't seem to be valid, though.

**Reply by nzakas** [reply]
This is consistent with how eslintrc normalized plugin names:

```yaml
plugins:
- @scope/foo/bar
```

ESLint would expand it to `@scope/eslint-plugin-foo/bar`. See [this test](https://github.com/eslint/eslintrc/blob/main/tests/lib/shared/naming.js#L25).

**Comment by meeroslav** [comment]
Feel free to change this from `bug` to `question`
